presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java

```diff
@@ -183,4 +189,21 @@ private boolean hasEnoughDiskSpace(Path path)
             throw new PrestoException(OUT_OF_SPILL_SPACE, "Cannot determine free space for spill", e);
         }
     }
+
+    private boolean isAccessible(Path path)
+    {
+        return isReadable(path) && isWritable(path) && isExecutable(path);
+    }
+
+    // Test whether a dir is working correctly by actually creating a random file in it
+    private boolean isSeeminglyHealthy(Path path)
+    {
+        try {
+            Path healthTemp = createTempFile(path, "spill", "healthcheck");
+            return deleteIfExists(healthTemp);
```

Comment:
This seems like a very expensive test to be doing for each file.  Instead, I suggest we have a Guava cache with a timeout to only perform the check every few seconds.

Reply:
This check in isolation is quite fast. It adds about 18-20% overhead on existing expensive check. Here is a benchmark result
```
Benchmark                                Mode  Cnt      Score      Error  Units
DiskHealthCheckBenchmark.combo          thrpt   10   8403.668 ±  152.380  ops/s
DiskHealthCheckBenchmark.spacebased     thrpt   10  10444.925 ±  122.971  ops/s
DiskHealthCheckBenchmark.tempfilebased  thrpt   10  80614.502 ± 1378.152  ops/s
```
Benchmark was
```  
@Benchmark
  public void spacebased(HealthCheckState state, Blackhole blackhole) throws Exception {
    FileStore store = Files.getFileStore(state.path);
    blackhole.consume(store.getTotalSpace() - store.getUsableSpace());
  }

  @Benchmark
  public void tempfilebased(HealthCheckState state, Blackhole blackhole) throws Exception {
    Path healthTemp = Files.createTempFile(state.path, "spill", "healthcheck");
    blackhole.consume(Files.deleteIfExists(healthTemp));
  }

  @Benchmark
  public void combo(HealthCheckState state, Blackhole blackhole) throws Exception {
    spacebased(state, blackhole);
    tempfilebased(state, blackhole);
  }
```

But I will yield to your suggestion as I am not familiar with the sensitivity to this additional overhead. Let me know your thoughts based on these numbers.

Reply:
The problem I see is that this is two extra disk operations.

Reply:
@dain Added a cache.